store/postgres/src/dynds/private.rs

```diff
@@ -96,10 +121,93 @@ impl DataSourcesTable {
         data_sources: &[StoredDynamicDataSource],
         block: BlockNumber,
     ) -> Result<usize, StoreError> {
-        todo!()
+        // Currently all data sources share the same causality region.
+        let causality_region = 0;
+
+        let mut inserted_total = 0;
+
+        for ds in data_sources {
+            let StoredDynamicDataSource {
+                manifest_idx,
+                param,
+                context,
+                creation_block,
+            } = ds;
+
+            if creation_block != &Some(block) {
+                return Err(constraint_violation!(
+                    "mismatching creation blocks `{:?}` and `{}`",
+                    creation_block,
+                    block
+                ));
+            }
+
+            let query = format!(
+                "insert into {}(block_range, manifest_idx, causality_region, param, context) \
+                 values (int4range($1, null), $2, $3, $4, $5)",
+                self.qname
+            );
+
+            inserted_total += sql_query(query)
+                .bind::<Nullable<Integer>, _>(creation_block)
+                .bind::<Integer, _>(*manifest_idx as i32)
+                .bind::<Integer, _>(causality_region)
+                .bind::<Nullable<Binary>, _>(param.as_ref().map(|p| &**p))
+                .bind::<Nullable<Jsonb>, _>(context)
+                .execute(conn)?;
+        }
+
+        Ok(inserted_total)
     }
 
     pub(crate) fn revert(&self, conn: &PgConnection, block: BlockNumber) -> Result<(), StoreError> {
-        todo!()
+        // Use `@>` to leverage the gist index.
+        // This assumes all ranges are of the form [x, +inf).
```

Comment:
If block ranges are always `[x, +inf)`, it would be better to not use a range at all, and just have a `block$` that stores `x` since that could be supported by a normal BTree index. But if we do need the flexibility to disable data sources afte ra certain block, this is ok.

Reply:
Yes this is trying to be future compatible with removing data sources by closing their range.